text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,570 @@
+## Summary
+
+Instead of every workspace package declaring a version range on a dependency, a new `workspace-consistent` version specifier protocol would allow packages to delegate to a range defined at the root `package.json`.
+
+## Motivation
+
+A common workflow in monorepos is the need to synchronize on the same version of a dependency.
+
+For example, the `foo` and `bar` packages of a monorepo may declare the same version of `react` in their `package.json` files.
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "^17.0.2"
+  }
+}
+```
+
+```json5
+// packages/bar/package.json
+{
+  "name": "@monorepo/bar",
+  "dependencies": {
+    "react": "^17.0.2"
+  }
+}
+```
+
+Multiple versions of the same dependency can cause different flavors of problems.
+
+- In projects that bundle dependencies, multiple versions inflate the size of the final result deployed to users.
+- Differing versions result in multiple copies that may not interact well at runtime, especially if features like [`Symbol()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are used. For example, [React hooks will error if a component is rendered with a different copy of React in the same app](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).
+- For TypeScript, multiple copies of the same `@types` package causes compile errors from mismatching definitions. The compiler diagnostic for this is usually: *"argument of type `Foo` is not assignable to parameter of type `Foo`"*. For developers that have seen this before, they may realize this diagnostic is due to a dependency version mismatch. For developers newer to TypeScript, *"`Foo` is not assignable to `Foo`"* is very confusing.
+
+While there are situations making differing versions become unavoidable, this is usually accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+## Detailed Explanation
+
+### Minimal Example
+
+<details open>
+<summary>A sample root `package.json` configuration would appear as such:</summary>
+
+```json5
+// package.json
+{
+  "name": "example-monorepo",
+  "private": true,
+  "pnpm": {
+    "workspaceConsistent": {
+      "versions": {
+        "jest": "^27.5.1",
+        "react": "^17.0.2",
+        "react-dom": "^17.0.2"
+      }
+    }
+  }
+}
+```
+</details>
+
+<details open>
+<summary>Workspace packages can then refer to the configured root defined version ranges:</summary>
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent",
```

Comment:
This syntax will not work. This means a dist-tag. The only solution is to use a new protocol
```suggestion
    "react": "workspace-consistent:",
```

Reply:
Thanks for catching! I'll update the doc for this later today.

A bit worried users will forget the `:` and see 404 errors for the dist tag. Maybe `workspace-consistent:default` should be required, but let's just do `workspace-consistent:` for now.

Reply:
Since it already needs to be a protocol, I would advocate for requiring a name after the protocol, and having that name refer to a defined "version set" of which multiple may be defined. The motivating example here is that a workspace may have a set of cutting edge projects configured to use `react@18`, while also having a set of legacy projects configured to use `react@17`, and the version set can be used to define the entire set of related packages (react-dom, react-test-renderer, etc.) and compatible versions.

The linked `syncpack` project already recognizes and supports this scenario. It's something I've been considering how to best support in [Rush](https://rushjs.io/), though with the constraint that the specified versions would only ever be allowed to be exact versions and it would use pnpmfile to instruct that any dependency of a dependency that can be satisfied by one of the listed versions must always do so (this is for compliance enforcement; allowing dependency versions to float has been repeatedly demonstrated of late to be a security issue waiting to happen).

Reply:
https://github.com/pnpm/rfcs/pull/1/commits/4cce21a3a766b12edeee47d4b2ca632e28fb55d5